### PR TITLE
Document asn.1 encoding functions

### PR DESCRIPTION
**Pull request by tashian**

### CODE REVIEW

File: step-ca/templates.mdx [hunk 1]
```diff
@@ -150,6 +152,95 @@ You can also import values
...
 in your provisioner's [configuration block](#configuring-the-ca-to-use-templates).
 
+### ASN.1 Values
+
+The X.509 templating system includes functions for ASN.1 encoding in a template.
+This can be useful for populating custom OID `extensions`:
+
+```
+"extensions": [
+  {"id": "1.2.3.4", "critical": true, "value": {{ asn1Enc "int:3848281" | toJson }}}
```
Comment (maraino) [inline step-ca/templates.mdx]: Usually critical will be `false` on custom extensions. If the client/server does not understand that extension it should fail.
```suggestion
  {"id": "1.2.3.4", "critical": false, "value": {{ asn1Enc "int:3848281" | toJson }}}
```

File: step-ca/templates.mdx [hunk 2]
```diff
@@ -150,6 +152,95 @@ You can also import values
...
+
+For example:
+
+```
+{"id": "1.2.3.4", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
+```
+
+1. `asn1Enc "YubiKey"` will encode "YubiKey" as an ASN.1 Printable string, and return the base64
+2. `asn1Enc "int:123456"` will encode 123 as an ASN.1 Integer, and return the base64
+3. `ans1Seq` wil get all the base64 strings, decode them, create an ASN.1 SEQENCE and return the base64 of the SEQUENCE.
```
Comment (maraino) [inline step-ca/templates.mdx]:
```suggestion
3. `asn1Seq` will get all the base64 strings, decode them, create an ASN.1 SEQENCE and return the base64 of the SEQUENCE.
```

File: step-ca/templates.mdx [hunk 3]
```diff
@@ -150,6 +152,95 @@ You can also import values
...
+4. Finally, `toJson` returns a JSON string representation of the base64
```
Comment (maraino) [inline step-ca/templates.mdx]: `asn1Set` is not described.

File: step-ca/templates.mdx [hunk 4]
```diff
@@ -150,6 +152,95 @@ You can also import values
...
+
+#### asn1Enc
+
+Encodes a string into an ASN.1 value.
+A data type can be supplied inline.
```
Comment (maraino) [inline step-ca/templates.mdx]: It supports the types described in [this function](https://github.com/smallstep/crypto/blob/878092791d7e5ef8530dfa949cfe3cde61373d77/x509util/extensions.go#L486):
* int: "int:1234"
* oid: "oid:1.2.3.4"
* utf8: "utf8:string"
* ia5: "ia5:srtring"
* numeric: "numeric:1234"
* printable: "printable:string" or just "string"
* utc: "utc:2023-03-29T02:03:57Z" or "utc:2023-04-20 15:28:27.909088 -0700 PDT"
* generalized: "generalized:2023-03-29T02:03:57Z" or "generalized:2023-04-20 15:28:27.909088 -0700 PDT"
* raw: a special one that allows any base64 string

File: step-ca/templates.mdx [hunk 5]
```diff
@@ -150,6 +152,95 @@ You can also import values
...
+
+`asn1Marshal` imitates Go's [`asn1.MarshalWithParams`](https://pkg.go.dev/encoding/asn1#MarshalWithParams).
+
+```
+asn1Marshal .Token.iss "printable"
+asn1Marshal .AuthorizationCrt.NotAfter "utc"
+```
+
+The second parameter ("printable", "utc") is the ASN.1 data type.
+The data types are listed in the [`encoding/asn1` package documentation](https://pkg.go.dev/encoding/asn1)
```
Comment (maraino) [inline step-ca/templates.mdx]: The second parameter is not required and it will use Go's default transformation.

**Review by maraino — CHANGES_REQUESTED** [review]

**Comment by tashian** [comment]
Hi @maraino could I get a final review on this one? Thanks

File: step-ca/templates.mdx [hunk 6]
```diff
@@ -150,7 +151,111 @@ You can also import values
...
+
+```
+{"id": "1.2.3.6", "value": {{ asn1Seq (asn1Enc "YubiKey") (asn1Enc "int:123456") | toJson }}}
+```
+
+To render this line of the template:
+
+1. `asn1Enc "YubiKey"` encodes "YubiKey" as an ASN.1 Printable string, and returns the Base64
+2. `asn1Enc "int:123456"` encodes 123 as an ASN.1 Integer, and returns the Base64
+3. `asn1Seq` gets the resulting Base64 strings, decodes them, creates an ASN.1 SEQENCE, and returns the Base64 of the SEQUENCE.
```
Comment (maraino) [inline step-ca/templates.mdx]:
```suggestion
3. `asn1Seq` gets the resulting Base64 strings, decodes them, creates an ASN.1 SEQUENCE, and returns the Base64 of the SEQUENCE.
```

File: step-ca/templates.mdx [hunk 7]
```diff
@@ -150,7 +151,111 @@ You can also import values
...
+```
+
+This function supports the data types described [here](https://github.com/smallstep/crypto/blob/878092791d7e5ef8530dfa949cfe3cde61373d77/x509util/extensions.go#L486):
+
+- **printable**: "printable:abcfoo" (or just "abcfoo")
+- **int**: "int:12324"
+- **oid**: "oid:1.2.3.4"
+- **utf8**: "utf8:foobar"
+- **ia5**: "ia5:abcfoo"
+- **numeric**: "numeric:01 02 03345" (multiple space-separated ints)
```
Comment (maraino) [inline step-ca/templates.mdx:219]: It allows numbers and spaces, you can also use numeric:1234, you don't have to use spaces.

**Review by maraino — CHANGES_REQUESTED** [review]
Just a typo

**Review by maraino — APPROVED** [review]

**Pull request merged**